wcfsetup/install/files/lib/http/middleware/Xsrf.class.php

```diff
@@ -34,11 +45,45 @@ public function process(ServerRequestInterface $request, RequestHandlerInterface
             self::TOKEN_ATTRIBUTE,
             $xsrfToken
         );
+
+        $hasValidXsrfToken = \hash_equals($xsrfToken, $request->getHeaderLine('x-xsrf-token'));
+
         $request = $request->withAttribute(
             self::HAS_VALID_HEADER_ATTRIBUTE,
-            \hash_equals($xsrfToken, $request->getHeaderLine('x-xsrf-token'))
+            $hasValidXsrfToken
         );
 
+        if (
+            !$this->isSafeHttpMethod($request->getMethod())
+            && $this->requestHandler->getActiveRequest()
+        ) {
+            $this->assertHasValidXsrfToken($this->requestHandler->getActiveRequest(), $hasValidXsrfToken);
+        }
+
         return $handler->handle($request);
     }
+
+    private function isSafeHttpMethod(string $verb): bool {
+        // HTTP requests using the 'GET' or 'HEAD' verb are safe
+        // by design, because those should not alter the state.
+        return $verb === 'GET' || $verb === 'HEAD';
+    }
+
+    private function assertHasValidXsrfToken(Request $request, bool $hasValidXsrfToken): void
+    {
+        if (!\is_subclass_of($request->getClassName(), RequestHandlerInterface::class)) {
+            // Skip the XSRF check for legacy controllers.
+            return;
+        }
+
+        $reflectionClass = new \ReflectionClass($request->getClassName());
+        if ($reflectionClass->getAttributes(DisableXsrfCheck::class) !== []) {
+            // Controller has opted out of the XSRF check.
+            return;
+        }
+
+        if (!$hasValidXsrfToken) {
+            throw new InvalidSecurityTokenException();
+        }
```

Comment:
```suggestion
        if ($hasValidXsrfToken) {
            // No need to do anything for a valid token.
            return;
        }

        if (!\is_subclass_of($request->getClassName(), RequestHandlerInterface::class)) {
            // Skip the XSRF check for legacy controllers.
            return;
        }

        $reflectionClass = new \ReflectionClass($request->getClassName());
        if ($reflectionClass->getAttributes(DisableXsrfCheck::class) !== []) {
            // Controller has opted out of the XSRF check.
            return;
        }

        // The controller requires a valid XSRF Token and no valid
        // token was provided, abort the processing.
        throw new InvalidSecurityTokenException();
```

One more optimization: Move the test for `$hasValidXsrfToken` to the top of this function and return early otherwise. This ensures that the reflection logic does not need to run for the happy-path of a valid token (which should be the majority of requests).